[CORE-8533] relax topic configs

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Fixes: #24543

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

### Bug Fixes

* Redpanda will now permit topics to be created with `redpanda.remote.[read|write]` set to `true` when a license is expired or missing provided that the cluster config `cloud_storage_enabled` is set to `false`.

michael-redpanda committed:
config: Added is_restricted() method to enterprise property

Signed-off-by: Michael Boquard <Michael@redpanda.com>

michael-redpanda committed:
kafka: Relaxed topic property enforcement when license invalid

Topic properties related to tiered storage and schema ID validation will
be permitted to be set/changed even when a license is not valid provided
that the overriding cluster config (`cloud_storage_enabled` and
`enable_schema_id_validation`) are disabled.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Test to verify relaxed tiered storage topic configs

Signed-off-by: Michael Boquard <michael@redpanda.com>

oleiman commented on src/v/kafka/server/tests/alter_config_test.cc:
```diff
@@ -234,6 +234,376 @@ class alter_config_test_fixture : public topic_properties_test_fixture {
+                  kafka::topic_property_record_key_subject_name_strategy,
+                  sns::topic_name)},
+              },
+              enable_cluster_config ? failure : success);
```
looks like `enable_cluster_config` is always true for these cases. is that intentional?

michael-redpanda replied:
Yes, I think I wrote the if statement after I put this in - i'll adjust

oleiman commented on src/v/kafka/server/tests/create_topics_test.cc:
```diff
@@ -545,6 +591,12 @@ FIXTURE_TEST(unlicensed_rejected, create_topic_fixture) {
 FIXTURE_TEST(unlicensed_reject_defaults, create_topic_fixture) {
+    lconf().cloud_storage_enabled.set_value(true);
+    auto unset_cloud_storage = ss::defer([&] {
+        lconf().enable_schema_id_validation.set_value(
+          pandaproxy::schema_registry::schema_id_validation_mode::none);
```
q: why is this needed?

michael-redpanda replied:
Probably just a copy-paste thing - I'll remove it

oleiman commented on tests/rptest/tests/license_enforcement_test.py:
```diff
@@ -201,3 +201,27 @@ def test_enterprise_cluster_bootstrap(self, root_driven_bootstrap):
+                                 err_msg="The cluster hasn't stabilized")
+
+        try:
+            self.rpk.create_topic("test",
+                                  config={"redpanda.remote.write": "true"})
+            assert not enable_cloud_storage, "Should have failed to create topic with redpanda.remote.write set and cloud_storage_enabled set to True"
```
is this the entirety of the issue? i vaguely understood that there was a question of whether upgrade was possible if 
- cloud_storage_enabled = false
- topics _already existing_ w/ remote_read or whatever set to true

Maybe I misunderstood though

michael-redpanda replied:
The upgrade should be possible - I can add in a test for that if you like

oleiman reviewed (dismissed):
generally lgtm. unfortunate that the unit test diff was so big. I went over it, but should probably do a second pass.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59742-->
<!--950c1d77-5a7c-47ba-8684-f8e9acaafab5-->
<!--506e1210-5f67-4246-ad32-c98ea91137dd-->
<!--title-->
## Retry command for [Build#59742](https://buildkite.com/redpanda/redpanda/builds/59742)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4b208aa5-60d0-4447-9868-01af83b328fc-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59742">build#59742</a></summary>


| test_id                                                                                                                                                                                                               | test_kind   | job_url                                                                                   | test_status   | passed   |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=20.cloud_storage_type=CloudStorageType.ABS                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59742#0193c201-a4ba-4b2b-964b-132c976ad079 | FAIL          | 0/6      |
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59742#0193c201-a4b7-4882-a1b4-07b07418b451 | FAIL          | 0/6      |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.S3                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59742#0193c201-a4ba-4b2b-964b-132c976ad079 | FLAKY         | 5/6      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59742#0193c201-a4b7-4882-a1b4-07b07418b451 | FAIL          | 0/1      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59742#0193c201-a4b7-4882-a1b4-07b07418b451 | FAIL          | 0/1      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59742#0193c201-a4b7-4882-a1b4-07b07418b451 | FAIL          | 0/1      |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59742#0193c201-a4b7-4882-a1b4-07b07418b451 | FAIL          | 0/1      |
</details>
<!--/body-->

IoannisRP reviewed (dismissed):
lgtm

BenPope commented on tests/rptest/tests/license_enforcement_test.py:
```diff
@@ -201,3 +201,27 @@ def test_enterprise_cluster_bootstrap(self, root_driven_bootstrap):
                                  backoff_sec=1,
                                  err_msg="The cluster hasn't stabilized")
+
+    @cluster(num_nodes=5)
```
5 nodes seems excessive.

michael-redpanda replied:
agreed - the test class actually uses 5 RP nodes, I can make a new class and use only three if you prefer

BenPope replied:
> agreed - the test class actually uses 5 RP nodes, I can make a new class and use only three if you prefer

Do any of the tests need 5 brokers?

michael-redpanda replied:
> > agreed - the test class actually uses 5 RP nodes, I can make a new class and use only three if you prefer
> 
> Do any of the tests need 5 brokers?

Not mine.  I didn't want to touch the original ones

BenPope reviewed (commented):
Makes sense to me.

michael-redpanda commented:
Force push [`98a08b2`]:

* Fixed logic in alter_config_test
* Removed unnecessary unsetting of enable_schema_id_validation in create_topics_test
* Added new test class that only uses 3 nodes
* Added test to verify that upgrades will work without a license, no TS enabled but topics with TS properties enabled exist

oleiman reviewed (approved):
lgtm

michael-redpanda commented:
No CI Failures: https://buildkite.com/redpanda/redpanda/builds/59806#0193cffc-aa95-43b3-a124-cb9b15923456 doesn't show any failed tests

vbotbuildovich commented:
/backport v24.3.x
